moveit_planners/ompl/ompl_interface/test/test_constrained_planning_state_space.cpp

```diff
@@ -0,0 +1,384 @@
+ * */
+class LoadTestRobotBaseClass : public testing::Test
+{
+protected:
+  LoadTestRobotBaseClass(const std::string& robot_name, const std::string& group_name)
+    : group_name_(group_name), robot_name_(robot_name)
+  {
+  }
+
+  void SetUp() override
+  {
+    // load robot
+    robot_model_ = moveit::core::loadTestingRobotModel(robot_name_);
+    robot_state_ = std::make_shared<robot_state::RobotState>(robot_model_);
+    joint_model_group_ = robot_state_->getJointModelGroup(group_name_);
+
+    // extract useful parameters for tests
+    num_dofs_ = joint_model_group_->getVariableCount();
+    ee_link_name_ = joint_model_group_->getLinkModelNames().back();
+    base_link_name_ = robot_model_->getRootLinkName();
+
+    // create a constrained state space for testing
+    setupMoveItStateSpace();
+    setupOMPLStateSpace();
+  };
+
+  void TearDown() override
+  {
+  }
+
+  Eigen::VectorXd getRandomState()
+  {
+    robot_state_->setToRandomPositions(joint_model_group_);
+    Eigen::VectorXd q;
+    robot_state_->copyJointGroupPositions(joint_model_group_, q);
+    return q;
+  }
+
+  /** \brief Create a joint position vector with values 0.1, 0.2, 0.3, ...
+   * where the length depends on the number of joints in the robot.
+   *
+   * We could also use a random state created in the method above,
+   * but do we want this randomness in tests??
+   * */
+  Eigen::VectorXd getDeterministicState()
+  {
+    Eigen::VectorXd state(num_dofs_);
+    double value = 0.1;
+    for (std::size_t i = 0; i < num_dofs_; ++i)
+    {
+      state[i] = value;
+      value += 0.1;
+    }
+    return state;
+  }
+
+  void setupMoveItStateSpace()
+  {
+    ompl_interface::ModelBasedStateSpaceSpecification space_spec(robot_model_, group_name_);
+    moveit_state_space_ = std::make_shared<ompl_interface::ConstrainedPlanningStateSpace>(space_spec);
+  }
+
+  void setupOMPLStateSpace()
+  {
+    // first call setupMoveItStateSpace()
+    assert(moveit_state_space_ != nullptr);
+    auto con = std::make_shared<DummyConstraint>(num_dofs_);
+    constrained_state_space_ = std::make_shared<ompl::base::ProjectedStateSpace>(moveit_state_space_, con);
+  }
+
+  void getValueAddressAtIndexTest()
+  {
+    Eigen::VectorXd joint_positions = getDeterministicState();
+    ompl::base::ScopedState<> ompl_state(constrained_state_space_);
+    auto state_ptr = ompl_state->as<ompl::base::ConstrainedStateSpace::StateType>()->getState();
+    double* out_joint_positions = dynamic_cast<ompl_interface::ModelBasedStateSpace::StateType*>(state_ptr)->values;
+    EXPECT_FALSE(out_joint_positions == nullptr);
+    for (std::size_t i = 0; i < num_dofs_; ++i)
+    {
+      *(out_joint_positions + i) = joint_positions[i];
+    }
+
+    // check getValueAddressAtIndex
+    for (std::size_t i = 0; i < num_dofs_; ++i)
+    {
+      EXPECT_EQ(joint_positions[i], *(moveit_state_space_->getValueAddressAtIndex(ompl_state.get(), i)));
```

Comment:
It better to have a tolerance when comparing. You can define a const value (called POSITION_TOLERANCE) for that and use it anywhere in the test:
``EXPECT_EQ(abs(joint_positions[i] - *(moveit_state_space_->getValueAddressAtIndex(ompl_state.get(), i))), POSITION_TOLERANCE);``

Reply:
Good point! I will change it. (The tolerance can be very small because it checks a value that was copied without any calculations, that's also why this exact equality worked probably.)